File: Medium/77. Combinations/combine.cpp
/*
    Time complexity: O(C(n, k))
    Space complexity: O(k)
*/
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ans;
        vector<int> cur;
        combination(n, k, ans, cur, 1);
        return ans;
    }
private:
    void combination(int n, int k, vector<vector<int>> &ans, vector<int> &cur, int s){
        if(cur.size() == k){
            ans.push_back(cur);
            return;
        }
        for(int i = s; i <= n; i++){
            cur.push_back(i);
            combination(n, k, ans, cur, i+1);
            cur.pop_back();
        }
    }
};

File: Medium/77. Combinations/combine.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combine(int n, int k) {
        ans = new ArrayList<List<Integer>>();
        List<Integer> track = new ArrayList<>();
        backtrack(n, k, 1, track);
        return ans;
    }
    private void backtrack(int n, int k, int start, List<Integer> track){
        if(k == track.size()){
            ans.add(new ArrayList<Integer>(track));
            return;
        }
        for(int i = start; i <= n; i++){
            track.add(i);
            backtrack(n, k, i + 1, track);
            track.remove(track.size() - 1);
        }
    }
}